fix(route): Anti-Anti-Spider 哔哩哔哩

NightSpaceC opened the pull request:
<!-- 
如有疑问，请参考 https://github.com/DIYgod/RSSHub/discussions/8002
Reference: https://github.com/DIYgod/RSSHub/discussions/8002
-->

## 该 PR 相关 Issue / Involved issue

Close #10511
#10823
#10824
#10831
#10898

## 完整路由地址 / Example for the proposed route(s)

<!--
请在 `routes` 区域填写以 / 开头的完整路由地址，否则你的 PR 将会被无条件关闭。
如果路由包含在文档中列出可以完全穷举的参数（例如分类），请依次全部列出。

Please include route starts with /, with all required and optional parameters. Fail to comply will result in your pull request being closed automatically.
```route
/some/route
/some/other/route
/dont/use/this/or/modify/it
/use/the/fenced/code/block/below
```
如果你的 PR 与路由无关, 请在 `routes` 区域 填写 `NOROUTE`，而不是直接删除 `routes` 区域。否则你的 PR 将会被无条件关闭。
If your changes are not related to route, please fill in `routes` with `NOROUTE`. Fail to comply will result in your PR being closed.
-->

```routes
/bilibili/user/video/335098315
/bilibili/user/video-all/335098315
```

## 新 RSS 检查列表 / New RSS Script Checklist
  
- [ ] 新的路由 New Route
  - [ ] 跟随 [v2 路由规范](https://docs.rsshub.app/joinus/script-standard.html) Follows [v2 Script Standard](https://docs.rsshub.app/en/joinus/script-standard.html)
- [ ] 文档说明 Documentation
  - [ ] 中文文档 CN
  - [ ] 英文文档 EN
- [ ] 全文获取 fulltext
  - [ ] 使用缓存 Use Cache
- [x] 反爬/频率限制 anti-bot or rate limit?
  - [x] 如果有, 是否有对应的措施? If yes, do your code reflect this sign?
- [ ] [日期和时间](https://docs.rsshub.app/joinus/pub-date.html) [date and time](https://docs.rsshub.app/en/joinus/pub-date.html)
  - [ ] 可以解析 Parsed
  - [ ] 时区调整 Correct TimeZone
- [ ] 添加了新的包 New package added
- [ ] `Puppeteer`

## 说明 / Note
这个 PR 试图通过在调用哔哩哔哩API之前访问`https://www.bilibili.com/`获得Cookie，并携带Cookie访问会使该API被调用的页面，最后携带Cookie访问API来反反爬虫

Commit:
fix(route): Anti-Anti-Reptilian 哔哩哔哩

Commit:
Merge branch 'master' of https://github.com/cyx20080216/RSSHub

Commit:
fix(route): Anti-Anti-Reptilian 哔哩哔哩

Commit:
fix(route): Remove debugging statement

Commit:
fix(route): Comply with code specifications by adding a semicolon.

sqliuchang commented:
实测这个PR可以正常抓取，运行了8小时了，没再出现用户名undefined了

TonyRL commented on lib/v2/bilibili/cache.js:
```diff
@@ -1,27 +1,75 @@
-        return ctx.cache.tryGet(key, async () => {
-            const { data: nameResponse } = await got(`https://api.bilibili.com/x/space/acc/info?mid=${uid}&jsonp=jsonp`, {
+    getCookie: async (ctx) => {
+        const key = 'cookie';
```
Please pick another cache key name like `bilibili:cookie`

TonyRL commented on lib/v2/bilibili/cache.js:
```diff
@@ -1,27 +1,75 @@
-            const { data: nameResponse } = await got(`https://api.bilibili.com/x/space/acc/info?mid=${uid}&jsonp=jsonp`, {
+    getCookie: async (ctx) => {
+        const key = 'cookie';
+        let cookie = await ctx.cache.get(key);
```
Please use `cache.tryGet`

TonyRL commented on lib/v2/bilibili/cache.js:
```diff
@@ -1,27 +1,75 @@
+        if (!cookie) {
+            const response = await got({
+                method: 'get',
+                url: `https://www.bilibili.com/`,
```
Why use template literals when there's no expression

TonyRL commented on lib/v2/bilibili/cache.js:
```diff
@@ -1,27 +1,75 @@
 const got = require('@/utils/got');
+const UA = require('@/utils/rand-user-agent')({ browser: 'chrome', os: 'windows', device: 'desktop' });
```
Does UA really matter since RSSHub is using `{ browser: 'chrome', os: 'mac os', device: 'desktop' }` by default

TonyRL commented on lib/v2/bilibili/cache.js:
```diff
@@ -1,27 +1,75 @@
+        const key = 'bili-username-from-uid-' + uid;
+        let name = await ctx.cache.get(key);
+        if (!name) {
+            await got({
+                method: 'get',
+                url: `https://space.bilibili.com/${uid}/`,
```
got request can be simplified
```suggestion
            await got(`https://space.bilibili.com/${uid}/`, {
```

TonyRL commented on lib/v2/bilibili/cache.js:
```diff
@@ -1,27 +1,75 @@
+    getUsernameFromUID: async (ctx, uid) => {
+        const cookie = await module.exports.getCookie(ctx);
+        const key = 'bili-username-from-uid-' + uid;
+        let name = await ctx.cache.get(key);
```
Please use `cache.tryGet`

TonyRL commented on lib/v2/bilibili/cache.js:
```diff
@@ -1,27 +1,75 @@
         let face = await ctx.cache.get(faceKey);
         if (!name) {
-            const nameResponse = await got(`https://api.bilibili.com/x/space/acc/info?mid=${uid}&jsonp=jsonp`, {
+            await got({
+                method: 'get',
+                url: `https://space.bilibili.com/${uid}/`,
```
```suggestion
            await got(`https://space.bilibili.com/${uid}/`, {
```

TonyRL commented on lib/v2/bilibili/video.js:
```diff
@@ -1,17 +1,30 @@
+        },
+    });
+    const response = await got({
+        method: 'get',
```
got request can be simplified

cyx20080216 commented on lib/v2/bilibili/video.js:
```diff
@@ -1,17 +1,30 @@
+    await got({
         method: 'get',
-        url: `https://api.bilibili.com/x/space/arc/search?mid=${uid}&ps=10&tid=0&pn=1&order=pubdate&jsonp=jsonp`,
+        url: `https://space.bilibili.com/${uid}/video?tid=0&page=1&keyword=&order=pubdate`,
```
I think we should use raw API format because browsers don't use the simplified format and it may become a reason to ban the spider.

TonyRL replied:
https://github.com/sindresorhus/got/blob/main/documentation/1-promise.md#promise-api

This is semantically **the same** as the first approach.

In fact, even your PR includes both <code>got(url: string | URL, options?: [OptionsInit](typescript.md#optionsinit), defaults?: [Options](2-options.md))</code> and <code>got(options: [OptionsInit](typescript.md#optionsinit))</code>

https://github.com/DIYgod/RSSHub/blob/de6dc376e38aba481f00efe98e6e8786d641d9d6/lib/v2/bilibili/video-all.js#L32-L38

https://github.com/DIYgod/RSSHub/blob/de6dc376e38aba481f00efe98e6e8786d641d9d6/lib/v2/bilibili/cache.js#L9-L15

cyx20080216 replied:
Sorry, you may misunderstand what I mean. I means the URL to call the API of Bilibili.

Commit:
fix(route): Resolve the requested changes
Use normative key
Use cache.tryGet
Use more simplified got request
Remove unnecessary template
Use default UA

Commit:
fix(route): Remove unnecessary await

Commit:
fix(route): Use simplified got request.

TonyRL commented on lib/v2/bilibili/cache.js:
```diff
@@ -1,30 +1,62 @@
+        return ctx.cache.tryGet(key, async () => {
+            const response = await got(`https://www.bilibili.com/`, {
+                headers: {
+                    'User-Agent': config.UA
```
No need to set UA again as it's already set in `lib/utils/got`
https://github.com/DIYgod/RSSHub/blob/3e6f0542c2ae51e9650376f4c435c109e548a85d/lib/utils/got.js#L33-L35

raining1123 commented:
请问更新到最新版本后，还需要手动在env里填写SESSDATA吗？

Commit:
fix(route): Use default UA

cyx20080216 commented:
> 请问更新到最新版本后，还需要手动在env里填写SESSDATA吗？

需要，这个Cookie是用于标识未登录用户身份的，不能获取您账户中的信息

Commit:
fix(route): Remove unnecessary require
